### DISCUSSION TITLE
Automatically perform 'SObject Definitions Refresh' after 'Retrieve Source from Org'

### DISCUSSION THREAD

**Discussion by SalesforceBobLightning** [forcedotcom/salesforcedx-vscode#5461] [Ideas]
I would like to suggest that the SalesforceDX VS Code extension automatically performs the 'SObject Definitions Refresh' command after the user performs the 'Retrieve Source from Org' command, because you need the objects refreshed to be able to develop against them.

What are the reasons for not doing this?

**Comment by renatoliveira** [comment]
I can think of some reasons it does not work this way today:

1. One does not necessarily need to refresh the definitions every time they retrieve the source from the organization.
2. It can take a long time on big projects. This is especially troublesome if people are still developing against a single sandbox (this is, AFAIK, pretty common for big pre-DX projects).

This does not mean I don't think this is a good idea. It really can help. But I think this can be **optional**. Maybe a `refreshDefinitionsOnRetrieve` flag on the project's configuration file.

**Comment by SalesforceBobLightning** [comment]
A flag sounds like a good solution.

What would the ETA on that be? :)

**Comment by lcampos** [comment]
Hi @SalesforceBobLightning, thanks for the feedback. I agree with you that this would make the overall experience better. Just to provide more context, the reason we didn't tackle it before were performance and priority. 

For performance, it's in line with what @renatoliveira already pointed out but we have some ideas on how to make it work in a more performant way. For priority, we wanted to focus on providing more value in the Apex LSP so at the time we focused on delivering Find All References, Go to Definition, etc. Now that we're working in Apex Rename Refactor this has come up internally as one of the things to tackle and it's currently in our queue. I'll make sure to update this thread once we have something close to be released.

**Comment by ntotten** [comment]
SObject refresh now happens on first load of a project. We don't pull on every refresh at this time, but its something being considered.

**Comment by j-isaacs** [comment]
I would propose you only refresh the definitions of SObjects included in the pull (CustomObject or their CustomFields). That would reduce the performance impact on pulls and greatly reduce if not eliminate the need for full SObject refreshes after the first load of a project.
